text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/StyleSerializer.java

```diff
@@ -277,6 +281,8 @@ public void write(final JsonWriter out, final Style value) throws IOException {
         this.gson.toJson(hoverEvent.value(), SerializerFactory.SHOW_ENTITY_TYPE, out);
       } else if (action == HoverEvent.Action.SHOW_TEXT) {
         this.gson.toJson(hoverEvent.value(), SerializerFactory.COMPONENT_TYPE, out);
+      } else if (action == HoverEvent.Action.SHOW_ACHIEVEMENT) {
+        this.gson.toJson(hoverEvent.value(), SerializerFactory.STRING_TYPE, out);
       } else {
         throw new JsonParseException("Don't know how to serialize " + hoverEvent.value());
       }
```

Comment:
I've been revisiting this a few times today (specifically in the context of [these lines](https://github.com/KyoriPowered/adventure/pull/890/files#diff-5109808b182eb56ebbaf1596841ba42362108e40c680c60a55b0f6ac568368bbR276-R292)) and I can't help but think that:
1. serialization of `"show_achievement"` should go in `serializeLegacyHoverEvent`, it does not belong in `"contents"` as during its lifetime it's only existed in `"value"`… which raises the question:
2. should the exception for "unknown hover actions" (Don't know how to serialize X) not be thrown when it's `"show_achievement"`? What should happen when `emitLegacyHover` is disabled too? Hover json serialization might need a review although that probably fits much better in a separate issue/discussion/pull request

This is mostly food for thought but I'd like to know the opinions from kashike and zml (and any other lurking living being!)

Reply:
I've pushed a change to emit the legacy hover event in *only* the legacy `value` field, rather than the modern `contents` field. Additionally, I've modified the serialization logic to just not serialize a hover event at all if legacy hover event output is not enabled and the action is `show_achievement`, just for efficiency's sake.